DaemonSet controller respects MaxSurge during update

smarterclayton opened the pull request:
If MaxSurge is set, the controller will attempt to launch updated pods on up to MaxSurge nodes and wait for them to be ready before deleting the old pods. If any old pod goes unready during update, a new pod is added to the node (regardless of MaxSurge) and the old pod is deleted.

This implements the logic behind #96375

/kind feature

```release-note
DaemonSets accept a MaxSurge integer or percent on their rolling update strategy that will launch the updated pod on nodes and wait for those pods to go ready before marking the old out-of-date pods as deleted. This allows workloads to avoid downtime during upgrades when deployed using DaemonSets. This feature is alpha and is behind the DaemonSetUpdateSurge feature gate.
```

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/1591-daemonset-surge
```

smarterclayton commented:
/sig apps

smarterclayton commented:
Added `DaemonSetUpdateSurge` to the e2e alpha suite in https://github.com/kubernetes/test-infra/pull/19914

smarterclayton commented:
/assign @kubernetes/sig-apps-pr-reviews

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/retest

smarterclayton commented:
The e2e test is in place, this is ready for an implementation review (I will be continuing to broaden the e2e test to disruption scenarios as the review continues).

kow3ns commented on test/e2e/apps/daemon_set.go:
```diff
@@ -482,6 +485,223 @@ var _ = SIGDescribe("Daemon set [Serial]", func() {
 		}
 	})
+
+	ginkgo.It("should surge pods onto nodes when spec was updated and update strategy is RollingUpdate [Feature:DaemonSetUpdateSurge]", func() {
```
Maybe add a note that we expect this to go to conformance after feature promotion? Sig-apps could be doing better on our conformance coverage and we're investigating. Having a reminder to update might be helpful.

smarterclayton replied:
Sure

kow3ns commented on pkg/controller/daemon/daemon_controller_test.go:
```diff
@@ -3186,3 +3193,62 @@ func getQueuedKeys(queue workqueue.RateLimitingInterface) []string {
+// Controller should not create pods on nodes which have daemon pods, and should remove excess pods from nodes that have extra pods.
+func TestSurgeDealsWithExistingPods(t *testing.T) {
+	ds := newDaemonSet("foo")
+	ds.Annotations = map[string]string{"alpha.apps.openshift.io/surge": "1"}
```
Maybe use something other than openshift for in tree k8s. Just a suggestion.

smarterclayton replied:
heh

smarterclayton replied:
The downsides of prototyping in a different fork

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented on pkg/controller/daemon/update.go:
```diff
@@ -415,12 +507,30 @@ func (dsc *DaemonSetsController) getUnavailableNumbers(ds *apps.DaemonSet, nodeL
+		}
+	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.DaemonSetUpdateSurge) {
+		if ds.Spec.UpdateStrategy.RollingUpdate.MaxSurge != nil {
```
I'm going to refactor this, I think it can be reduced cleanly.

smarterclayton commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -853,6 +855,39 @@ func (dsc *DaemonSetsController) podsShouldBeOnNode(
+		// If surge is allowed, there can be 2 pods on the node as long as:
+		// * the oldest new pod is not yet ready
+		// * the oldest old pod is ready
+		if util.AllowsSurge(ds) {
```
Yup, also cleaned it up a bit with comments

smarterclayton commented on pkg/controller/daemon/update.go:
```diff
@@ -415,12 +507,30 @@ func (dsc *DaemonSetsController) getUnavailableNumbers(ds *apps.DaemonSet, nodeL
+			return -1, -1, -1, fmt.Errorf("invalid value for MaxUnavailable: %v", err)
+		}
+	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.DaemonSetUpdateSurge) {
```
Brought this more cleanly in line with the util behavior as well

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
Adding some more testing, I'm seeing curious delays in DS deletion (which may be due to some transient errors occuring during DS deletePod calculation in 5 year old code).

Fixed a bug where we weren't waiting for MinReadySeconds for the new pod before deleting the old.  Added unit tests for that.

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
I'm going to restructure the update method a bit better - I think it may be possible to have rollingUpdate() only create on surge, and only delete on unavailable, which would be clearer and hide some failure modes that are ugly.

smarterclayton committed:
podutil: Use parenthesis for clarity around the pod ready condition

While this is correct in order of operations, it is harder to read
and masks the intent of the user without the parenthesis.

smarterclayton committed:
daemonset: Implement MaxSurge on daemonset update

If MaxSurge is set, the controller will attempt to double up nodes
up to the allowed limit with a new pod, and then when the most recent
(by hash) pod is ready, trigger deletion on the old pod. If the old
pod goes unready before the new pod is ready, the old pod is immediately
deleted. If an old pod goes unready before a new pod is placed on that
node, a new pod is immediately added for that node even past the MaxSurge
limit.

The backoff clock is used consistently throughout the daemonset controller
as an injectable clock for the purposes of testing.

smarterclayton committed:
daemonset: Simplify the logic for calculating unavailable pods

In order to maintain the correct invariants, the existing maxUnavailable
logic calculated the same data several times in different ways. Leverage
the simpler structure from maxSurge and calculate pod availability only
once, as well as perform only a single pass over all the pods in the
daemonset. This changed no behavior of the current controller, and
has a structure that is almost identical to maxSurge.

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton committed:
daemonset: Prevent accidental omissions of test condition checks

It is too easy to omit checking the return value for the
syncAndValidateDaemonSet test in large suites. Switch the method
type to be a test helper and fatal/error directly. Also rename
a method that referenced the old name 'Rollback' instead of
'RollingUpdate'.

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
Will push a verify fix after next review pass - this should be easier to read and better tested now (I caught a scenario where we waited longer than necessary to rollout the surge, so tried to organize the logic to actually match the simplest possible checks and let manage() handle the rest of the complextiy)

kow3ns reviewed (commented):
The algorithm looks correct as per the KEP. The logging needs to be cleaned up a bit. Some of the DEBUG logs should be removed. Others, I would argue, should have the DEBUG statement removed and be kept as they provide useful telemetry for failure conditions.

kow3ns commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -1290,6 +1333,14 @@ func (o podByCreationTimestampAndPhase) Less(i, j int) bool {
 	return o[i].CreationTimestamp.Before(&o[j].CreationTimestamp)
 }
 
+func podNames(pods []*v1.Pod) []string {
```
This doesn't appear to be used anywhere yet, consider deleting?

smarterclayton replied:
Yup, will be nuked

smarterclayton commented:
Yeah I plan to refresh all the debug logs into appropriate levels, just was keeping them and reviewing which were useful as I debugged prod issues.

smarterclayton committed:
daemonset: Remove unnecessary error returns in strategy code

The nodeShouldRunDaemonPod method does not need to return an error
because there are no scenarios under which it fails. Remove the
error return path for its direct calls as well.

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/retest
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/retest

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
/test pull-kubernetes-e2e-gce-alpha-features

smarterclayton commented:
Ok, this is ready for review (all known issues are addressed, the test is stable, logging should be accurate and at the right levels).  I'm fairly confident that this is in alpha shape at least.

jpbetz commented:
/remove-sig api-machinery

soltysh commented:
/triage accepted
/priority backlog

soltysh reviewed (approved):
This lgtm, leaving final tag to @kow3ns

smarterclayton commented:
/refresh

smarterclayton commented:
/refresh
/test pull-kubernetes-e2e-gce-alpha-features

soltysh reviewed (approved):
Having heard no objections for the past weeks, I'm tagging as is.
/lgtm
/retest
